crates/pallet-domains/src/block_tree.rs

```diff
@@ -0,0 +1,716 @@
+    }
+}
+
+/// Verify the execution receipt
+pub(crate) fn verify_execution_receipt<T: Config>(
+    domain_id: DomainId,
+    execution_receipt: &ExecutionReceiptOf<T>,
+) -> Result<(), Error> {
+    let ExecutionReceipt {
+        consensus_block_number,
+        consensus_block_hash,
+        domain_block_number,
+        block_extrinsics_roots,
+        parent_domain_block_receipt_hash,
+        ..
+    } = execution_receipt;
+
+    if domain_block_number.is_zero() {
+        // The genesis receipt is generated and added to the block tree by the runtime upon domain
+        // instantiation, thus it is unchallengeable and must always be the same.
+        ensure!(
+            BlockTree::<T>::get(domain_id, domain_block_number).contains(&execution_receipt.hash()),
+            Error::BadGenesisReceipt
+        );
+    } else {
+        let execution_inbox =
+            ExecutionInbox::<T>::get((domain_id, domain_block_number, consensus_block_number));
+        ensure!(
+            !block_extrinsics_roots.is_empty() && *block_extrinsics_roots == execution_inbox,
+            Error::InvalidExtrinsicsRoots
+        );
+    }
+
+    let excepted_consensus_block_hash =
+        frame_system::Pallet::<T>::block_hash(consensus_block_number);
+    ensure!(
+        *consensus_block_hash == excepted_consensus_block_hash,
+        Error::BuiltOnUnknownConsensusBlock
+    );
+
+    if let Some(parent_block_number) = domain_block_number.checked_sub(&One::one()) {
+        let parent_block_exist = BlockTree::<T>::get(domain_id, parent_block_number)
+            .contains(parent_domain_block_receipt_hash);
+        ensure!(parent_block_exist, Error::UnknownParentBlockReceipt);
+    }
+
+    match execution_receipt_type::<T>(domain_id, execution_receipt) {
+        ReceiptType::Rejected(RejectedReceiptType::InFuture) => {
+            log::error!(
+                "Unexpected in future receipt {execution_receipt:?}, which should result in \
+                `UnknownParentBlockReceipt` error as it parent receipt is missing"
+            );
+            Err(Error::InFutureReceipt)
+        }
+        ReceiptType::Rejected(RejectedReceiptType::Pruned) => {
+            log::error!(
+                "Unexpected pruned receipt {execution_receipt:?}, which should result in \
+                `InvalidExtrinsicsRoots` error as its `ExecutionInbox` is pruned at the same time"
+            );
+            Err(Error::PrunedReceipt)
+        }
+        ReceiptType::Accepted(_) | ReceiptType::Stale => Ok(()),
+    }
+}
+
+/// Process the execution receipt to add it to the block tree
+pub(crate) fn process_execution_receipt<T: Config>(
+    domain_id: DomainId,
+    submitter: OperatorId,
+    execution_receipt: ExecutionReceiptOf<T>,
+    receipt_type: AcceptedReceiptType,
+) -> Result<(), Error> {
+    let er_hash = execution_receipt.hash();
+    match receipt_type {
+        er_type @ AcceptedReceiptType::NewHead | er_type @ AcceptedReceiptType::NewBranch => {
```

Comment:
Having fewer branches looks more readable to me, non-blocker.

```rust
    let er_hash = execution_receipt.hash();
    let domain_block_number = execution_receipt.domain_block_number;
    let add_new_receipt_to_block_tree = || {
        // Construct and add a new domain block to the block tree
        let domain_block = DomainBlock {
            execution_receipt,
            operator_ids: sp_std::vec![submitter],
        };
        BlockTree::<T>::mutate(domain_id, domain_block_number, |er_hashes| {
            er_hashes.insert(er_hash);
        });
        DomainBlocks::<T>::insert(er_hash, domain_block);
    };
    match receipt_type {
        AcceptedReceiptType::NewBranch => add_new_receipt_to_block_tree(),
        AcceptedReceiptType::NewHead => {
            add_new_receipt_to_block_tree();

            // Update the head receipt number
            HeadReceiptNumber::<T>::insert(domain_id, domain_block_number);

            // Prune expired domain block
            if let Some(to_prune) =
                domain_block_number.checked_sub(&T::BlockTreePruningDepth::get())
            {
                for block in BlockTree::<T>::take(domain_id, to_prune) {
                    DomainBlocks::<T>::remove(block);
                }
                // Remove the block's `ExecutionInbox` as the block is pruned and does not need
                // to verify its receipt's `extrinsics_root` anymore
                let _ = ExecutionInbox::<T>::clear_prefix((domain_id, to_prune), u32::MAX, None);
            }
        }
        AcceptedReceiptType::CurrentHead => {
......
        }
    }
```

Reply:
Don't want to introduce a stateful closure here thus it is replaced by a function.